Character import speedup by delaying tag import

d-ber opened the pull request:
This PR significantly improves the speed of character bulk imports when dealing with a large number of characters by delaying tag import, partially addressing #2952.
Tag import requires fetching all characters via `getCharacters()`, which becomes expensive as the number of characters increases. 
Tag import is currently done after each character import, causing as many calls to `getCharacters()` as the number of characters imported.
When importing multiple characters, the import phase can be done after the character import, requiring only a single complete character update at the end.

## Checklist:

- [x] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

d-ber committed:
speed up char import by postponing tag import

Wolfsblvt reviewed (changes requested):
Generally, I am fine with this. This looks like a good performance improvement.
Very helpful.

Only thing I am having a bit trouble with is changing `importCharacter` to not **really** import the character, but having to call another method to import all data for it.
I know it's not an exported method, so it's only contained inside the code executed in this file.... *for now*.

Dunno if you understand where I'm coming from. Maybe there is a cleaner way to change this without ripping out a part of what "import character" means from the `importCharacter` function, so if any future dev who calls the method doesn't have to know to call XY method as well.

Maybe make the tag import optional as a parameter to the method, or add an `importCharacters` function that disables tag import for single import but imports all of them at the end, or something.

I'll think about something myself if you can't come up with something.

d-ber committed:
added tag import back into importCharacter

d-ber commented:
Added it back as a parameter which is set to false as default since both 2 current calls to the function don't need it. 
I implemented it as a call to the "new" function to avoid code duplication.

Wolfsblvt reviewed (approved):
Thank you for the implementing my suggested changes.
I am good with this now.

Cohee1207 commented on public/script.js:
```diff
@@ -8909,14 +8932,12 @@ async function importCharacter(file, preserveFileName = '') {
-            await importTags(importedCharacter);
+        let avatarFileName = `${data.file_name}.png`;
+        if (importTags) {
+            ImportCharactersTags([avatarFileName])
```
Shouldn't it be awaited? Especially since ImportCharactersTags uses getCharacters in it which could corrupt the app state if not careful.

Cohee1207 commented on public/script.js:
```diff
@@ -8851,24 +8851,47 @@ export async function processDroppedFiles(files, data = new Map()) {
+ */
+async function ImportCharactersTags(avatarFileNames) {
+    await getCharacters();
+    const currentContext = getContext();
```
Using `getContext` in these circumstances seems redundant. You can just query from characters array directly since they're declared in the same module.

Cohee1207 commented on public/script.js:
```diff
@@ -8851,24 +8851,47 @@ export async function processDroppedFiles(files, data = new Map()) {
+ * Imports tags for the given characters
+ * @param {string[]} avatarFileNames character avatar filenames whose tags are to import
+ */
+async function ImportCharactersTags(avatarFileNames) {
```
Please update to lowerCamelCase following naming conventions for introduced module functions.

Cohee1207 reviewed (changes requested):
Not a bad idea to batch tag imports together, but the code requires a little tiny bit of polishing.

d-ber committed:
fix function name, and minor fixes

d-ber committed:
fix toasts and final char selection

d-ber commented:
I fixed as you suggested. 
While testing, I noticed some issues with `select_rm_info()` being called before characters were updated, causing failed selections. I fixed it by simply showing the import toast for all but the last imported char, which gets selected too. `select_rm_info()` was adjusted to avoid a repeated import toast.

Cohee1207 commented on public/script.js:
```diff
@@ -7299,7 +7299,7 @@ export function select_rm_info(type, charId, previousCharId = null) {
     clearTimeout(importFlashTimeout);
     importFlashTimeout = setTimeout(function () {
-        if (type === 'char_import' || type === 'char_create') {
+        if (type === 'char_import' || type === 'char_create' || type === 'char_import_no_toast') {
```
I don't see `char_import_no_toast` being used anywhere else and it appears redundant.

Cohee1207 replied:
Resolved.

Cohee1207 committed:
chore: reformat new code

Cohee1207 reviewed (approved):
Everything works for me. Tested individual, bulk imports, drag&drops.

ObscenityIB commented:
Fantastic!

I am glad the old speed is back, after the recent update that drastically slowed down character imports 10x or more, I had little hope of the bug being fixed as it was "technically functional" and therefore not a priority.
